### PR TITLE
Added Strassen's Algorithm and Weighted Job Scheduling.

### LINKED ISSUE #12: UVa Maximum Sub-sequence Product
Use **Java BigInteger Library** to code this DP [problem](https://uva.onlinejudge.org/index.php?option=com_onlinejudge&Itemid=8&category=649&page=show_problem&problem=728).

### PR DESCRIPTION
**Pull request by syedjahangirpeeran**
Fixes #12 .

### CODE REVIEW

**Comment by syedjahangirpeeran** [comment]
@sahilbansal17  review my code as part of GSSoC.

**Comment by syedjahangirpeeran** [comment]
Fixes #11 Weighted Job Scheduling.

**Comment by syedjahangirpeeran** [comment]
@sahilbansal17  Please go through it so that I can start solving others issues.

**Comment by syedjahangirpeeran** [comment]
Done

**Comment by syedjahangirpeeran** [comment]
Added #38 .

File: Algorithms/Divide-and-Conquer/Strassens Algorithm.cpp [hunk 1]
```diff
@@ -0,0 +1,129 @@
+#include <bits/stdc++.h>
+#include <vector>
+#include<stdio.h>
+#define ll long long int
+using namespace std;
+void sumsub(vector< vector<ll> > &A, vector< vector<ll> > &B, vector< vector<ll> > &C, ll m, char s) {
```
Comment (sahilbansal17) [inline Algorithms/Divide-and-Conquer/Strassens Algorithm.cpp]: Please write appropriate comments what each function does and also follow the naming conventions as mentioned in the [Coding standards wiki page](https://github.com/sahilbansal17/Competitive_Coding/wiki/Coding-Standards)

File: Algorithms/Divide-and-Conquer/Strassens Algorithm.cpp [hunk 2]
```diff
@@ -0,0 +1,129 @@
+#include <bits/stdc++.h>
+#include <vector>
```
Comment (sahilbansal17) [inline Algorithms/Divide-and-Conquer/Strassens Algorithm.cpp]: No need of additional header files, if you have already included "bits/stdc++.h"

File: Algorithms/Divide-and-Conquer/Strassens Algorithm.cpp [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+            for(j=0;j<n;j++){
+                cout<<C[i][j]<<" ";
+            }
+            cout<<endl;
+        }
+
+    }
+return 0;
+}
+//From Master's Theorem, time complexity is O(N^Log7) which is approximately O(N^2.8074)
```
Comment (sahilbansal17) [inline Algorithms/Divide-and-Conquer/Strassens Algorithm.cpp]: Please elaborate more on this part, what is the recurrence relation and how exactly it came up.

File: Algorithms/Divide-and-Conquer/Strassens Algorithm.cpp [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+        sumsub(p1, p3, A1, s, '+');
+        sumsub(A1, p6, B1, s, '+');
+        sumsub(B1, p2, c22, s, '-');
+ 
+        join(c11, c12, c21, c22, C ,s);
+    }
+    
+int main(){
+    ll t ,i ,j ,k, q ,n;
+    cin>>t;
```
Comment (sahilbansal17) [inline Algorithms/Divide-and-Conquer/Strassens Algorithm.cpp]: Please enter appropriate cout statements as prompt for the user to know what is being taken as input!

File: Algorithms/Divide-and-Conquer/Strassens Algorithm.cpp [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+    for(k=0;k<t;k++){
+        scanf("%lld",&n);
+        vector<ll> x (n);
+        vector< vector<ll> > A(n, x), B(n, x), C(n, x);
+        
+        for(i=0;i<n;i++){
+            for(j=0;j<n;j++){
+                scanf("%lld",&q);
+                A[i][j]=q;
```
Comment (sahilbansal17) [inline Algorithms/Divide-and-Conquer/Strassens Algorithm.cpp]: Follow the [Coding standards](https://github.com/sahilbansal17/Competitive_Coding/wiki/Coding-Standards) with appropriate spacing.

File: Algorithms/Divide-and-Conquer/Strassens Algorithm.cpp [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+        sumsub(B1, p2, c22, s, '-');
+ 
+        join(c11, c12, c21, c22, C ,s);
+    }
+    
+int main(){
+    ll t ,i ,j ,k, q ,n;
+    cin>>t;
+    for(k=0;k<t;k++){
+        scanf("%lld",&n);
```
Comment (sahilbansal17) [inline Algorithms/Divide-and-Conquer/Strassens Algorithm.cpp]: Do not use both **cin** and **scanf** statements in the same program. Use either of them.

File: Algorithms/Divide-and-Conquer/Weighted Job Scheduling.cpp [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+
+struct Job{
+    int start, finish, profit;
+};
+
+bool compareFinishTime(Job s1, Job s2){
+    return (s2.finish > s1.finish);
+}
+
+int binarySearch(Job jobs[], int index){// to find the latest job (before current job) that doesn't conflict with current job.
```
Comment (sahilbansal17) [inline Algorithms/Divide-and-Conquer/Weighted Job Scheduling.cpp]: Please follow to [Coding standards](https://github.com/sahilbansal17/Competitive_Coding/wiki/Coding-Standards) for comments. Comments should be above the code statements.

File: Algorithms/Divide-and-Conquer/Weighted Job Scheduling.cpp [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+    }
+
+    result = profitTill[n-1];      // Store result and free dynamic memory allocated for profitTill[]
+    delete[] profitTill;
+
+    return result;
+}
+
+int main(){
+    Job array[] = {{1, 2, 150}, {5, 10, 80}, {3, 11, 20}, {2, 50, 300}};
```
Comment (sahilbansal17) [inline Algorithms/Divide-and-Conquer/Weighted Job Scheduling.cpp]: Take input from user with appropriate prompts using **cout**.

File: Algorithms/Divide-and-Conquer/Weighted Job Scheduling.cpp [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+    int n = sizeof(array)/sizeof(array[0]);
+    cout << "Maximum profit obtained is " << findMaxProfit(array, n);
+    return 0;
+}
+// The complexity of this algorithm is O(nLogn) as it uses Binary Search.
```
Comment (sahilbansal17) [inline Algorithms/Divide-and-Conquer/Weighted Job Scheduling.cpp]: Elaborate on the complexity. Add atleast a few lines making it more clear.

**Comment by sahilbansal17** [comment]
@syedjahangirpeeran Please make the requested changes.

**Comment by sahilbansal17** [comment]
@syedjahangirpeeran This PR would be closed if there is no longer any response from your side.

**Pull request closed without merging**